docs/HowToFAQ/Errors-Parser.md

```diff
@@ -0,0 +1,1194 @@
+import opened N
+module {:options opt} M{}
+```
+
+The value of an options attribute cannot be a computed expression. It must be a literal string.
+
+## **Error: cannot declare identifier beginning with underscore**
+
+```dafny
+const _myconst := 5
+function m(): (_: int) {0}
+```
+
+User-declared identifiers may not begin with an underscore; 
+such identifiers are reserved for internal use. 
+In match statements and expressions, an 
+identifier that is a single underscore is used as a wild-card match.
+
+<!-- There are two instances of this message. An example of the other message is
+     function m(): (_: int) {0}
+-->
+
+## **Error: sorry, bitvectors that wide (_number_) are not supported**
+
+```dafny
+const b: bv2147483648
+```
+
+A bitvector type name is the characters 'bv' followed by a non-negative
+integer literal. However, dafny only supports widths up to the largest
+signed 32-bit integer (2147483647).
+
+## **Error: sorry, arrays of that many dimensions (_number_) are not supported**
+
+```dafny
+const a: array2147483648<int>
+```
+
+A multi-dimensional array type name is the characters 'array' followed by
+a positive integer literal. However, dafny only supports numbers of 
+dimensions up to the largest signed 32-bit integer (2147483647).
+In practice (as of v3.9.1) dimensions of more than a few can take 
+overly long to resolve ([Issue #3180](https://github.com/dafny-lang/dafny/issues/3180)).
+
+## **Error: There is no point to an export declaration at the top level**
+
+```dafny
+export M
+method M() {}
+```
+
+Although all top-level declarations are contained in an implicit top-level module, there is no syntax to import that module.
+Such an import would likely cause a circular module dependency error.
+If the module cannot be imported, there is no point to any export declarations.
+
+## **Error: expected either a '{' or a 'refines' keyword here, found _token_**
+
+```dafny
+module M {}
+module N refine M {}
+```
+
+The [syntax for a module declaration](../DafnyRef/DafnyRef#sec-modules) is either `module M { ... }` or
+`module M refines N { ... }` with optional attributes after the `module` keyword.
+This error message often occurs if the `refines` keyword is misspelled.
+
+## **Error: no comma is allowed between provides and reveals and extends clauses in an export declaration**
+
+<!-- There are four instances of this error message -->
+
+```dafny
+module M {
+  export A reveals b, a, reveals b
+  export B reveals b, a, provides b
+  export C provides b, a, reveals b
+  export D provides b, a, provides b
+  const a: int
+  const b: int
+}
+```
+
+An export declaration consists of one or more `reveals`, `provides`, and extends clauses. Each clause contains
+a comma-separated list of identifiers, but the clauses themselves are not separated by any delimiter.
+So in the example above, the comma after `a` is wrong in each export declaration. 
+This mistake is easy to make when the clauses are on the same line.
```

Comment:
The comma ought to be allowed (and obviously optional). It reads so much better in English with a comma. No need to take action on that, just me being *grumph* :-)

Reply:
OK. And Rustan voted against it being optional. And that would cause some cascading ambiguities.